reference-implementation/lib/writable-stream.js

```diff
@@ -182,215 +190,226 @@ function WritableStreamAddWriteRequest(stream) {
-  }
+function WritableStreamFinishInFlightWriteWithError(stream, reason) {
+  assert(stream._inFlightWriteRequest !== undefined);
+  stream._inFlightWriteRequest._reject(reason);
+  stream._inFlightWriteRequest = undefined;
 
-  if (state === 'errored') {
-    WritableStreamRejectPromisesInReactionToError(stream);
+  const state = stream._state;
 
+  if (state === 'errored') {
+    WritableStreamFinishInFlightWriteInErroredState(stream);
     return;
   }
 
+  assert(state === 'writable');
+
   stream._state = 'errored';
   stream._storedError = reason;
 
-  if (wasAborted === false && stream._writer !== undefined) {
-    WritableStreamDefaultWriterEnsureReadyPromiseRejectedWith(stream._writer, reason, readyPromiseIsPending);
+  if (stream._pendingAbortRequest === undefined) {
+    const writer = stream._writer;
+    if (writer !== undefined) {
+      if (WritableStreamCloseQueuedOrInFlight(stream) === false && stream._backpressure === true) {
+        defaultWriterReadyPromiseReject(writer, reason);
+      } else {
+        defaultWriterReadyPromiseResetToRejected(writer, reason);
+      }
+      writer._readyPromise.catch(() => {});
+    }
+  } else {
+    stream._pendingAbortRequest._reject(reason);
+    stream._pendingAbortRequest = undefined;
   }
 
-  WritableStreamRejectPromisesInReactionToError(stream);
+  WritableStreamRejectPromisesInReactionToError(stream, reason);
 }
 
-function WritableStreamFinishPendingClose(stream) {
-  assert(stream._pendingCloseRequest !== undefined);
-  stream._pendingCloseRequest._resolve(undefined);
-  stream._pendingCloseRequest = undefined;
+function WritableStreamFinishInFlightClose(stream) {
+  assert(stream._inFlightCloseRequest !== undefined);
+  stream._inFlightCloseRequest._resolve(undefined);
+  stream._inFlightCloseRequest = undefined;
 
   const state = stream._state;
 
-  let wasAborted = false;
-  if (stream._pendingAbortRequest !== undefined) {
-    wasAborted = true;
-  }
-
   if (state === 'errored') {
-    if (wasAborted === true) {
-      stream._pendingAbortRequest._reject(stream._storedError);
-      stream._pendingAbortRequest = undefined;
-    }
-
-    WritableStreamRejectClosedPromiseIfAny(stream);
-
+    WritableStreamFinishInFlightCloseInErroredState(stream);
     return;
   }
 
-  assert(state === 'closing');
+  assert(state === 'writable');
+
+  if (stream._pendingAbortRequest === undefined) {
+    stream._state = 'closed';
 
-  if (wasAborted === false) {
     const writer = stream._writer;
     if (writer !== undefined) {
       defaultWriterClosedPromiseResolve(writer);
     }
-    stream._state = 'closed';
+
     return;
   }
 
   stream._pendingAbortRequest._resolve();
   stream._pendingAbortRequest = undefined;
 
+  const error = new TypeError('Closed after abort request');
+
   stream._state = 'errored';
-  stream._storedError = new TypeError('Abort requested but closed successfully');
+  stream._storedError = error;
+
+  // TODO(tyoshino): No need to reject readyPromise?
```

Comment:
Sorry. Never mind. It seems I was confused.

Abort()-ing a stream with in-flight close resets the readyPromise to a rejected promise, and it's final for the readyPromise as well as other places where we reject (or reset-to-rejected) the readyPromise. So, here we don't have to do anything for the readyPromise.